### PR TITLE
Annotate System.Net.Security for nullable reference types

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to #2339

I think I might have spotted a few bugs, please see my comments in the diff /cc @dotnet/ncl

### CODE REVIEW

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ public enum EncryptionPolicy
         AllowNoEncryption = 1,
         NoEncryption = 2,
     }
-    public delegate System.Security.Cryptography.X509Certificates.X509Certificate LocalCertificateSelectionCallback(object sender, string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection localCertificates, System.Security.Cryptography.X509Certificates.X509Certificate remoteCertificate, string[] acceptableIssuers);
+    public delegate System.Security.Cryptography.X509Certificates.X509Certificate LocalCertificateSelectionCallback(object sender, string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection localCertificates, System.Security.Cryptography.X509Certificates.X509Certificate? remoteCertificate, string[] acceptableIssuers);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs]: I presume `sender` should be declared nullable here?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 2]
```diff
@@ -184,28 +184,28 @@ private SecurityStatusPal DecryptData()
...
 
             if (isAsync)
             {
-                result = ForceAuthenticationAsync(new AsyncSslIOAdapter(this, cancellationToken), _context.IsServer, null, isApm);
+                result = ForceAuthenticationAsync(new AsyncSslIOAdapter(this, cancellationToken), _context!.IsServer, null, isApm);
             }
             else
             {
-                ForceAuthenticationAsync(new SyncSslIOAdapter(this), _context.IsServer, null).GetAwaiter().GetResult();
+                ForceAuthenticationAsync(new SyncSslIOAdapter(this), _context!.IsServer, null).GetAwaiter().GetResult();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: This `Task?` result permeates all the way up to the public API. I would recommend changing so that `Task.CompletedTask` is returned in the synchronous branch.

Unless intentional, we should additionally also make this method async and await the authentication method.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 3]
```diff
@@ -440,7 +440,7 @@ private int CheckEnqueueRead(Memory<byte> buffer)
                 _lockReadState = LockPendingRead;
             }
             // Need to exit from lock before waiting.
-            lazyResult.InternalWaitForCompletion();
+            lazyResult!.InternalWaitForCompletion();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I might be missing something, but I don't see how this can be non-null.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 4]
```diff
@@ -521,7 +521,7 @@ private void CheckEnqueueWrite()
             }
 
             // Need to exit from lock before waiting.
-            lazyResult.InternalWaitForCompletion();
+            lazyResult!.InternalWaitForCompletion();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Same here. @wfurt?

File: src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs [hunk 5]
```diff
@@ -31,7 +31,7 @@ public IEnumerable<TlsCipherSuite> AllowedCipherSuites
             {
                 // This method is only useful only for diagnostic purposes so perf is not important
                 // We do not want users to be able to cast result to something they can modify
-                foreach (TlsCipherSuite cs in Pal.GetCipherSuites())
+                foreach (TlsCipherSuite cs in Pal.GetCipherSuites()!)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs]: `Pal.GetCipherSuites()` always returns null.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: > This Task? result permeates all the way up to the public API.

What do you mean?  The result Task is only ever used when isAsync is true.

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: > Unless intentional, we should additionally also make this method async and await the authentication method.

This is intentional.  This method should not be made async.

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Do we ever invoke this will null as the sender?  Are there any situations where any code you've found passes null?  Is there code that creates these delegates not defensively checking that sender is null?

File: src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs]: >  always returns null

Did you mean non-null?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: > What do you mean? The result Task is only ever used when isAsync is true. This is not observable by callers of the public method.

You're right. I suppose though we could change line 198 to `Task result = Task.CompletedTask;` in order to avoid having to make nullability assertions based on input arguments.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Addressed in https://github.com/dotnet/runtime/pull/32541/commits/a613646eb341dc4e0b1158a294f86a706d35eda6

File: src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs (hunk 5 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs]: Null, at least in the case of windows https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicyPal.Windows.cs#L16

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/CipherSuitesPolicy.cs]: > Null, at least in the case of windows

We will never get here on Windows because the ctor will throw a PlatformNotSupportedException.

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs (hunk 1 above)
Comment (iSazonov) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: See https://source.dot.net/#System.Net.Security/System/Net/Security/SecureChannel.cs,655
Here `string.Empty` is used as `sender`.

See https://source.dot.net/#System.Net.Security/System/Net/Security/SecureChannel.cs,360
Here TargetHost is used as `sender`. Default for the TargetHost is string.Empty https://source.dot.net/#System.Net.Security/System/Net/Security/SslAuthenticationOptions.cs,47

Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Non-nullable it is then.

**Comment by eiriktsarpalis** [comment]
Rebased on top of newly added System.Net.Primitives annotations

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 6]
```diff
@@ -54,34 +54,34 @@ public partial class NegotiateStream : System.Net.Security.AuthenticatedStream
...
+        public virtual System.IAsyncResult BeginAuthenticateAsClient(System.Net.NetworkCredential credential, string targetName, System.AsyncCallback asyncCallback, object? asyncState) { throw null; }
+        public virtual System.IAsyncResult BeginAuthenticateAsClient(System.Net.NetworkCredential credential, string targetName, System.Net.Security.ProtectionLevel requiredProtectionLevel, System.Security.Principal.TokenImpersonationLevel allowedImpersonationLevel, System.AsyncCallback asyncCallback, object? asyncState) { throw null; }
+        public virtual System.IAsyncResult BeginAuthenticateAsServer(System.AsyncCallback asyncCallback, object? asyncState) { throw null; }
+        public virtual System.IAsyncResult BeginAuthenticateAsServer(System.Net.NetworkCredential credential, System.Net.Security.ProtectionLevel requiredProtectionLevel, System.Security.Principal.TokenImpersonationLevel requiredImpersonationLevel, System.AsyncCallback? asyncCallback, object? asyncState) { throw null; }
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Some `System.AsyncCallback? asyncCallback` are nullable in this list, and some aren't. For Sockets, I made all `BeingXXX` methods have `System.AsyncCallback? asyncCallback, object? asyncState` parameters, as that seemed to be the established pattern (ex. Stream.BeginRead/Write).

Comment (stephentoub) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Which ones aren't nullable? That's almost certainly a mistake if there are any. All AsyncCallback and object state instances in the APM pattern can be null.

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 7]
```diff
@@ -197,13 +197,13 @@ public partial class SslStream : System.Net.Security.AuthenticatedStream
...
+        public virtual System.IAsyncResult BeginAuthenticateAsClient(string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection? clientCertificates, bool checkCertificateRevocation, System.AsyncCallback asyncCallback, object asyncState) { throw null; }
+        public virtual System.IAsyncResult BeginAuthenticateAsClient(string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection? clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation, System.AsyncCallback asyncCallback, object asyncState) { throw null; }
         public virtual System.IAsyncResult BeginAuthenticateAsServer(System.Security.Cryptography.X509Certificates.X509Certificate serverCertificate, System.AsyncCallback asyncCallback, object asyncState) { throw null; }
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs]: `object asyncState` => `object? asyncState`.

Also, same comment here as above on `AsyncCallback`.

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 8]
```diff
@@ -222,7 +222,7 @@ public partial class SslStream : System.Net.Security.AuthenticatedStream
         public void Write(byte[] buffer) { }
         public override void Write(byte[] buffer, int offset, int count) { }
         public override System.Threading.Tasks.Task WriteAsync(byte[] buffer, int offset, int count, System.Threading.CancellationToken cancellationToken) { throw null; }
-        public override System.Threading.Tasks.ValueTask WriteAsync(System.ReadOnlyMemory<byte> buffer, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public override System.Threading.Tasks.ValueTask WriteAsync(System.ReadOnlyMemory<byte> buffer, System.Threading.CancellationToken cancellationToken) { throw null; }
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Was this change intended?

**Comment by eerhardt** [comment]
>         public InvalidCredentialException(string message, System.Exception innerException) { }

`string? message, System.Exception? innerException`.  Here and for `string message` overload above.

---
Refers to: src/libraries/System.Net.Security/ref/System.Net.Security.cs:583 in cadecbf. [](commit_id = cadecbf680f43be03eb4fd7ca59df230ffe075ba, deletion_comment = False)

**Comment by eerhardt** [comment]
>         public bool Contains(string searchServiceName) { throw null; }

From a quick look at the code, it appears that `Contains(string searchServiceName)` handles passing in `null`. Should this be `string? searchServiceName`?

---
Refers to: src/libraries/System.Net.Security/ref/System.Net.Security.cs:617 in cadecbf. [](commit_id = cadecbf680f43be03eb4fd7ca59df230ffe075ba, deletion_comment = False)

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 9]
```diff
@@ -591,13 +591,13 @@ public partial class ExtendedProtectionPolicy : System.Runtime.Serialization.ISe
         public ExtendedProtectionPolicy(System.Security.Authentication.ExtendedProtection.PolicyEnforcement policyEnforcement) { }
         public ExtendedProtectionPolicy(System.Security.Authentication.ExtendedProtection.PolicyEnforcement policyEnforcement, System.Security.Authentication.ExtendedProtection.ChannelBinding customChannelBinding) { }
         public ExtendedProtectionPolicy(System.Security.Authentication.ExtendedProtection.PolicyEnforcement policyEnforcement, System.Security.Authentication.ExtendedProtection.ProtectionScenario protectionScenario, System.Collections.ICollection customServiceNames) { }
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs]: `System.Collections.ICollection customServiceNames`, the parameter appears to allow null.

[https://github.com/dotnet/runtime/blob/8ddc3e11454b1f4cd4df27003d1bbfbb1212e9d1/src/libraries/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicy.cs#L42-L48](https://github.com/dotnet/runtime/blob/8ddc3e11454b1f4cd4df27003d1bbfbb1212e9d1/src/libraries/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ExtendedProtectionPolicy.cs#L42-L48)

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Crypto.cs [hunk 10]
```diff
@@ -87,12 +87,12 @@ internal static int BioTell(SafeBioHandle bio)
 
         internal static string? GetX509RootStorePath()
         {
-            return Marshal.PtrToStringAnsi(GetX509RootStorePath_private());
+            return Marshal.PtrToStringAnsi(GetX509RootStorePath_private())!;
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Crypto.cs]: Why do you need these two changes? `GetX509RootStorePath` and `GetX509RootStoreFile` return `string?`.

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs [hunk 11]
```diff
@@ -151,7 +152,7 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
 
                 if (hasCertificateAndKey)
                 {
-                    SetSslCertificate(innerContext, certHandle, certKeyHandle);
+                    SetSslCertificate(innerContext, certHandle!, certKeyHandle!);
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: I'm surprised Roslyn can't figure this out themselves. We check `!= null` above and store it in a bool.

Comment (stephentoub) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: https://github.com/dotnet/roslyn/issues/34665
https://github.com/dotnet/csharplang/issues/2388

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs [hunk 12]
```diff
@@ -135,7 +136,7 @@ internal static byte[] SslGetAlpnSelected(SafeSslHandle ssl)
 
         internal static string GetOpenSslCipherSuiteName(SafeSslHandle ssl, TlsCipherSuite cipherSuite, out bool isTls12OrLower)
         {
-            string ret = Marshal.PtrToStringAnsi(GetOpenSslCipherSuiteName(ssl, (int)cipherSuite, out int isTls12OrLowerInt));
+            string ret = Marshal.PtrToStringAnsi(GetOpenSslCipherSuiteName(ssl, (int)cipherSuite, out int isTls12OrLowerInt))!;
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: Looking at the [C implementation](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c#L383), `CryptoNative_GetOpenSslCipherSuiteName` can return `null`. And the one call site of this method handles getting `null` back. So maybe this method should return `string?`.

File: src/libraries/Common/src/Interop/Windows/SspiCli/NegotiationInfoClass.cs [hunk 13]
```diff
@@ -31,7 +32,7 @@ internal static string GetAuthenticationPackageName(SafeHandle safeHandle, int n
                     string name;
                     unsafe
                     {
-                        name = Marshal.PtrToStringUni(((SecurityPackageInfo*)packageInfo)->Name);
+                        name = Marshal.PtrToStringUni(((SecurityPackageInfo*)packageInfo)->Name)!;
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Windows/SspiCli/NegotiationInfoClass.cs]: If we declared `string name` as `string? name`, you could get rid of this suppression.

File: src/libraries/Common/src/Interop/Windows/SspiCli/ISSPIInterface.cs [hunk 14]
```diff
@@ -10,22 +11,22 @@ namespace System.Net
...
         int EncryptMessage(SafeDeleteContext context, ref Interop.SspiCli.SecBufferDesc inputOutput, uint sequenceNumber);
         int DecryptMessage(SafeDeleteContext context, ref Interop.SspiCli.SecBufferDesc inputOutput, uint sequenceNumber);
         int MakeSignature(SafeDeleteContext context, ref Interop.SspiCli.SecBufferDesc inputOutput, uint sequenceNumber);
         int VerifySignature(SafeDeleteContext context, ref Interop.SspiCli.SecBufferDesc inputOutput, uint sequenceNumber);
 
         int QueryContextChannelBinding(SafeDeleteContext phContext, Interop.SspiCli.ContextAttribute attribute, out SafeFreeContextBufferChannelBinding refHandle);
-        int QueryContextAttributes(SafeDeleteContext phContext, Interop.SspiCli.ContextAttribute attribute, Span<byte> buffer, Type handleType, out SafeHandle refHandle);
+        int QueryContextAttributes(SafeDeleteContext? phContext, Interop.SspiCli.ContextAttribute attribute, Span<byte> buffer, Type? handleType, out SafeHandle? refHandle);
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Windows/SspiCli/ISSPIInterface.cs]: Both places that implement this `QueryContextAttributes` interface method just suppress the nullable warning on `phContext`. Why not change the interface to say it doesn't accept `null` for `phContext`?

File: src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs [hunk 15]
```diff
@@ -453,13 +457,13 @@ public static string QueryStringContextAttributes(ISSPIInterface secModule, Safe
                     return null;
                 }
 
-                string result = Marshal.PtrToStringUni(sspiHandle.DangerousGetHandle());
+                string result = Marshal.PtrToStringUni(sspiHandle.DangerousGetHandle())!;
```
Comment (eerhardt) [inline src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs]: The `QueryStringContextAttributes` method returns `string?`, why not just declare `string? result` and remove the suppression?

File: src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs [hunk 16]
```diff
@@ -15,8 +16,8 @@ namespace System.Net.Security
 {
     internal sealed class SafeFreeSslCredentials : SafeFreeCredentials
     {
-        private SafeX509Handle _certHandle;
-        private SafeEvpPKeyHandle _certKeyHandle;
+        private SafeX509Handle _certHandle = null!;
```
Comment (eerhardt) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs]: Both of these fields can be `null` if this object was created with `certificate == null`. Shouldn't these be nullable fields?

File: src/libraries/Common/src/System/Net/NTAuthentication.Common.cs [hunk 17]
```diff
@@ -46,7 +47,7 @@ internal string ClientSpecifiedSpn
                     _clientSpecifiedSpn = GetClientSpecifiedSpn();
                 }
 
-                return _clientSpecifiedSpn;
+                return _clientSpecifiedSpn!;
```
Comment (eerhardt) [inline src/libraries/Common/src/System/Net/NTAuthentication.Common.cs]: `GetClientSpecifiedSpn()` can return `null`. Are we sure this will never be null?

Maybe we can change `NegotiateStreamPal.QueryContextClientSpecifiedSpn` to return a `string` and not a nullable string?

File: src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs [hunk 18]
```diff
@@ -233,7 +233,7 @@ internal void Write(ReadOnlySpan<byte> buf)
 
         internal int BytesReadyForConnection => _toConnection.Count;
 
-        internal byte[] ReadPendingWrites()
+        internal byte[]? ReadPendingWrites()
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs]: This looks like it always returns a non-nullable `byte[]`. Why this change?

File: src/libraries/System.Net.Security/src/System/Net/Security/NegoState.cs [hunk 19]
```diff
@@ -579,10 +581,10 @@ private void StartReceiveBlob(LazyAsyncResult lazyResult)
                 message = _framer.EndReadMessage(ar);
             }
 
-            ProcessReceivedBlob(message, lazyResult);
+            ProcessReceivedBlob(message!, lazyResult);
         }
 
-        private void ProcessReceivedBlob(byte[] message, LazyAsyncResult lazyResult)
+        private void ProcessReceivedBlob(byte[] message, LazyAsyncResult? lazyResult)
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/NegoState.cs]: For this one, I think it should be fine to say `byte[]? message` here, and not need to suppress the callsites. It appears to be "normal flow" to pass null into here in order to throw an AuthenticationException.

File: src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 20]
```diff
@@ -218,11 +218,11 @@ private X509Certificate2 EnsurePrivateKey(X509Certificate certificate)
                 }
 
                 X509Certificate2Collection collectionEx;
-                string certHash = certEx.Thumbprint;
+                string certHash = certEx!.Thumbprint;
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: If `MakeEx` returns null above, this will NRE. Is this a bug?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs [hunk 21]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs]: Are these still needed if they aren't in the `Common` folder?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 22]
```diff
@@ -184,28 +184,29 @@ private SecurityStatusPal DecryptData()
...
             if (isAsync)
             {
-                result = ForceAuthenticationAsync(new AsyncSslIOAdapter(this, cancellationToken), _context.IsServer, null, isApm);
+                result = ForceAuthenticationAsync(new AsyncSslIOAdapter(this, cancellationToken), _context!.IsServer, null, isApm);
             }
             else
             {
-                ForceAuthenticationAsync(new SyncSslIOAdapter(this), _context.IsServer, null).GetAwaiter().GetResult();
+                ForceAuthenticationAsync(new SyncSslIOAdapter(this), _context!.IsServer, null).GetAwaiter().GetResult();
+                result = null;
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: For my understanding here - why not change this to `result = Task.CompletedTask` and change the return type to be `Task` instead of `Task?`? In the synchronous case, the return is ignored. And it saves from having to suppress the null-check up stream.

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: Hmm, the implementation seems to not accept a default value so this was reflected by the reference source generator. I'm not sure whether the discrepancy is a something that needs fixing, so I'll just revert this.

File: src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs [hunk 23]
```diff
@@ -24,8 +24,9 @@ static TlsCipherSuiteData()
                 s_tlsLookup.Count == LookupCount,
                 $"Lookup dictionary was of size {s_tlsLookup.Count} instead of {LookupCount}");
 
-            foreach (TlsCipherSuite val in Enum.GetValues(typeof(TlsCipherSuite)))
+            foreach (object? value in Enum.GetValues(typeof(TlsCipherSuite)))
```
Comment (eerhardt) [inline src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: This doesn't seem like a nice pattern. Is this really how you have to enumerate enum values now?

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: https://github.com/dotnet/csharplang/issues/3214

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs (hunk 6 above)
Comment (eerhardt) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs]: > Which ones aren't nullable?

The 3 directly above this line aren't nullable, for example.

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Crypto.cs (hunk 10 above)
Comment (eiriktsarpalis) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Crypto.cs]: Looks like I added it initially and then forgot to remove. Thanks.

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs (hunk 11 above)
Comment (eiriktsarpalis) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: I presume that's intentional though. There's only so many scenaria you can support  before complexity goes out of hand.

File: src/libraries/Common/src/Interop/Windows/SspiCli/ISSPIInterface.cs (hunk 14 above)
Comment (eiriktsarpalis) [reply src/libraries/Common/src/Interop/Windows/SspiCli/ISSPIInterface.cs]: There are quite a few call sites passing potentially null values to this method. I've updated the signature and forced the value on the call sites instead.

File: src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs (hunk 16 above)
Comment (eiriktsarpalis) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs]: Good catch, thanks.

File: src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs (hunk 20 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs]: It seems likely. Maybe somebody from @dotnet/ncl could provide context?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs (hunk 21 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs]: Many implementation files are referenced by the unit test project directly, which is not annotated. This is one of those files.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs (hunk 22 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Had a long conversation with @stephentoub about this already, relevant thread: https://github.com/dotnet/runtime/pull/32541#discussion_r381320794

File: src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs (hunk 21 above)
Comment (eerhardt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslApplicationProtocol.cs]: Makes sense. Thanks.

**Comment by eiriktsarpalis** [comment]
Rebased on top of System.Net.Sockets annotations.

File: src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs (hunk 23 above)
Comment (eerhardt) [reply src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: I almost want to suggest to shutting off `#nullable` around this loop, that's how disappointed I am with this result.

**Review by eerhardt — APPROVED** [review]
Assuming your questions to the library owners are resolved, this LGTM. Nice work.

File: src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs (hunk 23 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: That's what dotnet/csharplang#3214 is about fixing.

Comment (eerhardt) [reply src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: I meant in the meantime, until it is fixed in the compiler.

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs]: We should annotate things the way everyone else will need to.  I don't think we'll advocate wrapping every foreach in a #nullable disable.  Either this gets fixed in the compiler, or we may need to make the `IEnumerator` interface itself oblivious.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs (hunk 3 above)
Comment (iSazonov) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: What was a conclusion here? Looks tricky if it is not a bug.

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: We never get here.  It's a known bug. @wfurt has been working on it.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Tracking in #32920

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: https://github.com/dotnet/runtime/pull/32925
